features/steps/tabs.py

```diff
@@ -0,0 +1,79 @@
+# encoding: utf-8
+
+"""
+Step implementations for paragraph-related features
+"""
+
+from behave import given, then, when
+
+from docx import Document
+from docx.text.tabstop import TabStop
+from docx.shared import Inches
+
+from helpers import test_docx
+
+
+# given ===================================================
+@given('a tab_stops having {count} tab stops')
+def given_a_tab_stops_having_count_tab_stops(context, count):
+    count_int = {
+        '0': 0,
+        '3': 1
+    }[count]
+    document = Document(test_docx('tab-stops'))
+    paragraph_format = document.paragraphs[count_int].paragraph_format
+    context.tab_stops = paragraph_format.tab_stops
+
+
+# when ====================================================
+
+@when('I add a tab_stop')
+def when_i_add_a_tab_stop(context):
+    tab_stops = context.tab_stops
+    tab = tab_stops.add_tab()
+    tab.position = Inches(1.75)
+
+
+@when('I delete a tab_stop')
+def when_i_delete_a_tab_stop(context):
+    tab_stops = context.tab_stops
+    tab = tab_stops[1]
+    tab.delete()
```

Comment:
I can see wanting to eliminate a tab when working with a whole set of 
tab_stops, but I can also imagine working with an individual tab and 
wanting to delete it.  For example, if I'm working with an individual 
tab to change it's position, but move it to a position where another tab 
already exists, I might want to delete at the tab level.

I've got it implemented, and don't see any problematic behaviors. It's 
behaving exactly as I want it to.  What sort of problems should I look for?

I don't care if it's called delete() or remove(), as far as Python 
method names go, but we should be consistent across levels.

David



On 04/20/2016 12:39 PM, Steve Canny wrote:
>
> In features/steps/tabs.py 
> <https://github.com/python-openxml/python-docx/pull/261#discussion_r60453707>:
>
> > +
> > +
> > +# when ====================================================
> > +
> > +@when('I add a tab_stop')
> > +def when_i_add_a_tab_stop(context):
> > +    tab_stops = context.tab_stops
> > +    tab = tab_stops.add_tab()
> > +    tab.position = Inches(1.75)
> > +
> > +
> > +@when('I delete a tab_stop')
> > +def when_i_delete_a_tab_stop(context):
> > +    tab_stops = context.tab_stops
> > +    tab = tab_stops[1]
> > +    tab.delete()
>
> Can you help me understand your reasoning?
>
> The behavior of TabStops.remove(tab_stop) is to remove tab_stop as a 
> member of the TabStops sequence. So it seems natural to me that the 
> call is on TabStops. After tab_stop is removed, it would no longer 
> have a parent and so a .remove() call on it would be invalid.
>
> So I'm not seeing what we would get by having both, and there would be 
> additional complexity to account for if we did. Also, I'd like to 
> reuse base Python method names whenever possible to minimize 
> surprising behavior.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/python-openxml/python-docx/pull/261/files/dbfce4c2185a8718c44be6ddd260be5c5e6856ff..98ba6b5433023e9d8d3d32330d089063cad8e1bd#r60453707>
>